### ISSUE TITLE
Lambda Monitoring - Draft

### ISSUE THREAD

**Issue by cartersocha** [open-telemetry/opentelemetry-specification#3042]
### Description

_Describe the goals, objectives, and requirements for the project. This include the motivations for starting the project now, as opposed to later._

The goal of this project is to put Lambda monitoring into a consistently good state. Across vendors, customers consistently struggle to instrument their Lambdas and identify the best practices way to monitor a Lambda. Lambda layer behavior differs by language, context propagation is frequently broken, and cold starts are a known issue. 

We want to ensure:

-  Consistent lambda layer behavior by language and uniform conformance to the spec
-  There is clear community guidance on how to monitor Lambdas 
- End to end context propagation using community protocols / propagation across typical Lambda architectures (e.g async)
- Cold start data capture and submission
- The new TelemetryAPI is properly integrated and utilized by OpenTelemetry

### Project Board

_Once approved by TC, a project should be managed using a GitHub project board. This project board should be pre-populated with issues that cover all known deliverables, organized by timeline milestones.

A [Technical Committee](https://github.com/open-telemetry/community/blob/main/community-members.md#technical-committee) (TC) member associated with the project can create the board, along with a new project-specific GitHub label to automatically associate issues and PRs with the project. The project lead and all other relevant project members should have edit access to the board.

Once created, please link to the project board here._

### Deliverables

_A description of what this project is planning to deliver, or is in the process of delivering. This includes all OTEPs and their associated prototypes.

In general, OTEPs are not accepted unless they come with working prototypes available to review in at least two languages. Please discuss these requirements with a TC member before submitting an OTEP._

### Staffing / Help Wanted

_Who is currently planning to work on the project? If a project requires specialized domain expertise, please list it here. If a project is missing a critical mass of people in order to begin work, please clarify._

Lightstep, AWS, and Honeycomb are interested in improving this area. 

While Lambdas are the focus of this effort we need **other Functions As A Service (FAAS) experts** to ensure we're building conventions that make sense for the stateless function space in general. GCP or Azure participation would be welcome. 

#### Required staffing

_Projects cannot be started until they the following participants have been identified:
* Every project needs a project lead, who is willing to bottom line the project and address any issues which are not handled by other project members.
* At least two sponsoring TC members. TC sponsors are dedicated to attending meetings, reviewing proposals, and in general being aware of the state of the project and it's technical details. TC sponsors guide the project through the spec process, keep the tracking issue up to date, and help to ensure that relevant community members provide input at the appropriate times.
* Engineers willing to write prototypes in at least two languages (if relevant to project). Languages should be fairly different from each other (for example: Java and Python).
* Maintainers or approvers from those languages committed to reviewing the prototypes._

Project Lead: TBA
Sponsoring TC Members: TBA
Implementation Engineers: TBA
Implementation Maintainers or Approvers: Matt Wear and Diego Hurtado (Placeholders)
Functions SME(s): 

### Meeting Times

_Once a project is started, the working group should meet regularly for discussion. These meeting times should be posted on the OpenTelemetry public calendar._

To deliver the improvements promptly we propose meeting at least 2 days a week for the 6 week planning cycle as specified in the new [Semantic Conventions Process Doc](https://docs.google.com/document/d/1ghvajKaipiNZso3fDtyNxU7x1zx0_Eyd02OGpMGEpLE/edit#heading=h.xhmejuvvum3i)

TBD on when. The current (dead) SIG invite is set for Wednesdays at 4 PM PST. 

### Timeline

_What is the expected timeline the project will aim to adhere to, and what resources and deliverables will be needed for each portion of the timeline? If the project has not been started, please describe this timeline in relative terms (one month in, two weeks later, etc). If a project has started, please include actual dates._

1.  New working group will be kicked off in January 
2.  The WG has 6 weeks to propose improvements to the specification and solutions  -  Mid to Late February
3.  Implementation - we want to start with JavaScript and Python as our first target implementation languages - End of April (Placeholder implementation timeline)

### Labels

_The tracking issue should be properly labeled to indicate what parts of the specification it is focused on._

### Linked Issues and PRs

_All PRs, Issues, and OTEPs related to the project should link back to the tracking issue, so that they can be easily found._

**Comment by svrnm** [comment]
👍

**Comment by arbiv** [comment]
We (Cisco, formerly Epsagon) are also interested in improving this area and willing to participate in the engineering efforts in JS and Python. 

We already started to contribute the documentation (https://github.com/open-telemetry/opentelemetry.io/pull/1974).

**Comment by cartersocha** [comment]
@codeboten, @tylerbenson - please add any known lambda deficiencies into issues and link here for the final issue section or send to me over Slack.

**Comment by cartersocha** [comment]
> 👍
> 
> I would love seeing this getting some traction. There is plenty of tutorials out there that describe how you do serverless (AWS, Azur, GCP) with OpenTelemetry and each of them has their specific way of doing it, so it's super confusing for endusers what the right way of doing it is.
> 
> > There is clear community guidance on how to monitor Lambdas
> 
> We started to have some initial documentation on serverless for JS on AWS in the official docs (https://opentelemetry.io/docs/instrumentation/js/serverless/), and plan to have more (e.g. here's a PR for GCP [open-telemetry/opentelemetry.io#2091](https://github.com/open-telemetry/opentelemetry.io/pull/2091) and a list where we track the effort ([open-telemetry/opentelemetry.io#2021](https://github.com/open-telemetry/opentelemetry.io/issues/2021))

+ @lmolkova if there's an interest in Azure Functions

**Comment by tigrannajaryan** [comment]
Is this proposal and https://github.com/open-telemetry/opentelemetry-lambda repo related in any way?

**Comment by cartersocha** [comment]
@tigrannajaryan yep! The restarted SIG would take over the repo but not sure the exact approver / maintainer list. 
@Aneurysm9 will be overall lead

**Comment by niko-achilles** [comment]
> We want to ensure:
> Consistent lambda layer behavior by language and uniform conformance to the spec
> There is clear community guidance on how to monitor Lambdas
> End to end context propagation using community protocols / propagation across typical Lambda architectures (e.g async)
> Cold start data capture and submission
> The new TelemetryAPI is properly integrated and utilized by OpenTelemetry

I like that @cartersocha .  One question the registry could also be a point of improvement  ? when i search for aws in context of lambda in registry, i get confused to select the right instrumentation

**Comment by svrnm** [comment]
@cartersocha I think it would be good to rename this to "serverless" instead of "lambda" where possible.

**Comment by cartersocha** [comment]
@svrnm agreed there is a need to address stateless more broadly.  

However with the current state of lambda instrumentation, customer experience, and previous community support I’d like to keep this effort focused on lambdas. I’m not sure if the lambda repo even has an owner right now besides Anthony occasionally looking at it but customers are trying to use the packages in production. 

A subsequent or parallel effort would be needed with function representatives from multiple vendors for the full stateless scope.

**Comment by tylerbenson** [comment]
@cartersocha I created an issue describing the biggest issue I see: #3060

**Comment by svrnm** [comment]
> However with the current state of lambda instrumentation, customer experience, and previous community support I’d like to keep this effort focused on lambdas

Although I agree that with putting a focus on lambda first, I think naming the WG / effort "serverless" is more inclusive, so it also invites ppl that want to pick up other kinds of faas & serverless into the work. It's similar to "Client Side Telemetry" which is not named "Browser RUM" or "Mobile RUM", although they have a strong focus on that matter.

**Comment by Aneurysm9** [comment]
> > However with the current state of lambda instrumentation, customer experience, and previous community support I’d like to keep this effort focused on lambdas
> 
> Although I agree that with putting a focus on lambda first, I think naming the WG / effort "serverless" is more inclusive, so it also invites ppl that want to pick up other kinds of faas & serverless into the work. It's similar to "Client Side Telemetry" which is not named "Browser RUM" or "Mobile RUM", although they have a strong focus on that matter.

I agree with this, mostly.  I think the WG should definitely be "FaaS" but may still encompass activities that are directly and solely targeted at Lambda.  For instance, the current https://github.com/open-telemetry/opentelemetry-lambda repo provides a Lambda Extension and Layers that, to the best of my knowledge, do not have analogues in other FaaS offerings and that, should any such analogous offering exist, may still require independent implementations.  We should seek to make specification as broad as possible but retain the flexibility to provide more focused implementations as necessary.

I believe FaaS is the appropriate scope as "serverless" is in the eye of the beholder.

**Comment by cartersocha** [comment]
Updated the item and issue title based on the feedback @Aneurysm9, @svrnm. Initial SIG focus will be on the fixing the current lambda state but the SIG will broadly cover the FAAS space. 

Agreed on the spec being broadly applicable.

**Comment by nozik** [comment]
I suggest making request/response hooks a standard capability for serverless integrations (already implemented in Node, I've created a PR in Python - [link](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1476)). In several languages, request/response hooks are provided for HTTP clients and servers, allowing the extraction of data that isn't collected  by default. A typical use case for that would be extracting a customer id from a header/body parameter, and adding it as an attribute to the span, to be later on used for filtering spans by customer. For that matter, serverless functions aren't much different than HTTP endpoints - and in some cases, like connecting AWS API Gateway to a Lambda, they are logically acting as HTTP servers.

Generally speaking, these hooks are an important enabler for adapting OTEL to each engineer organization's specific needs - and also help downstream distros innovate in various domains based on the core OTEL SDK.

**Comment by ocelotl** [comment]
I asked Claude to review this issue and Claude suggests it be closed with this explanation:

> A project-tracking meta-issue for a FaaS/Lambda working group kicked off Dec 2022; last activity was Jan 17 2023 (nozik's comment) and the effort never produced further tracking updates here. Maintainer cartersocha noted 'I'm unsure if more spec work is needed for FAAS. Our initial findings have mostly been implementation inconsistencies,' and the actual deliverables (FaaS semconv, Lambda layers, cold-start processor) live in the semantic-conventions, opentelemetry-lambda, and opentelemetry.io repos rather than this spec repo. A long-dead SIG tracking issue with deliverables tracked elsewhere.

To guard against a mistaken suggestion, Claude then independently re-verified this conclusion:

> Confirmed: FAAS SIG project-tracking meta-issue, dead since Jan 17 2023 (3.5+ yrs); maintainer cartersocha questioned whether spec work is even needed, deliverables live in opentelemetry-lambda/semantic-conventions/opentelemetry.io, and both linked issues (#3060, community #685) are closed. Stale-abandoned holds.

Claude evaluated this issue with medium confidence that it can be closed.

Claude has classified this issue as stale.

The intention of this comment is to make it easier for a human being to review this issue before making a decision.

**Comment by mx-psi** [comment]
Closing this given this is a completed project: https://github.com/open-telemetry/community/blob/main/projects/completed-projects/faas.md
